### PR TITLE
Support expression in f-strings spec

### PR DESCRIPTION
**Pull request by palaviv**
This allows:
```py
spec = "0>+#10x"
f"{16:{spec}}"
```

### CODE REVIEW

File: parser/src/fstring.rs [hunk 1]
```diff
@@ -209,6 +245,42 @@ mod tests {
...
+        let parse_ast = parse_fstring(&source).unwrap();
+
+        assert_eq!(
+            parse_ast,
+            FormattedValue {
+                value: Box::new(mk_ident("foo", 1, 1)),
+                conversion: None,
+                spec: Some(Box::new(Constant {
+                    value: "spec".to_string(),
+                })),
```
Comment (coolreader18) [inline parser/src/fstring.rs]: Could you add some more tests for some weirder cases? e.g. `f"{val: {'#'} }"` interprets the spec as `' {\'#\'} '`; it only interpolates the value if the only characters for the spec are `:{expression}}`

File: parser/src/fstring.rs [hunk 2]
```diff
@@ -43,13 +43,49 @@ impl<'a> FStringParser<'a> {
...
-                        if next != '}' {
-                            spec.push(next);
-                            self.chars.next();
-                        } else {
-                            break;
+                        match next {
+                            '{' => {
+                                if in_nested {
+                                    return Err(ExpressionNestedTooDeeply);
+                                } else {
```
Comment (coolreader18) [inline parser/src/fstring.rs]: You don't need this `else`, as you return in the first condition.

File: tests/snippets/fstrings.py [hunk 3]
```diff
@@ -16,6 +16,9 @@
 assert f'{16:0>+#10x}' == '00000+0x10'
 assert f"{{{(lambda x: f'hello, {x}')('world}')}" == '{hello, world}'
 
+spec = "0>+#10x"
+assert f"{16:{spec}}{foo}" == '00000+0x10bar'
+
```
Comment (coolreader18) [inline tests/snippets/fstrings.py]: Could you add some TypeError tests here for formatting with `int`s as specs or similar?

Comment (palaviv) [reply tests/snippets/fstrings.py]: Could you be more specific?

File: parser/src/fstring.rs (hunk 2 above)
Comment (palaviv) [reply parser/src/fstring.rs]: Done

File: parser/src/fstring.rs (hunk 1 above)
Comment (palaviv) [reply parser/src/fstring.rs]: I know this is not perfect yet but we can improve on that later. I added a few more fail cases tests.

File: tests/snippets/fstrings.py (hunk 3 above)
Comment (coolreader18) [reply tests/snippets/fstrings.py]: e.g. `assert_raises(TypeError, lambda: f"{123:{45}}")`

Comment (palaviv) [reply tests/snippets/fstrings.py]: > e.g. `assert_raises(TypeError, lambda: f"{123:{45}}")`

This is actually a valid format string.
I added a TODO in the tests for an invalid one. This currently does not fail but it is not related to this change.

**Comment by doyshinda** [comment]
Hey, I have a PR for supporting `:f` string format codes in #1651 and I noticed that after rebasing with these changes, one of my new tests is now failing with `f-string` style formatting:
```Python
assert f'{5.1234: f}' == ' 5.123400'
```

When using the `f` style formatting, the `" "` prefix is missing:
```Python
Welcome to the magnificent Rust Python 0.1.1 interpreter 😱 🖖
>>>>> f'{5.1234: f}'
'5.123400'
>>>>> '{: f}'.format(5.1234)
' 5.123400'
>>>>> '{x: f}'.format_map({'x': 5.1234})
' 5.123400'
>>>>>
```

**Comment by coolreader18** [comment]
It looks like that came from line 86 in `parser/sec/fstring.rs`:

```rust
spec = Some(Box::new(Constant {
    value: spec_expression.trim().to_string(),
}))
```

If you remove the `.trim()` in your branch, that should fix it.

**Comment by doyshinda** [comment]
Confirmed, removing `trim()` fixed it, thanks!

**Pull request merged**